File: langs/csharp/src/VMx/Components/ComponentVM.cs
using VMx.Builders;
using VMx.Services;

namespace VMx.Components;

/// <summary>
/// Sealed, non-modeled leaf viewmodel.
///
/// See spec/05-component-vm.md §Variants (ComponentVM, type=Component). Use
/// <c>ComponentVM.Builder()</c> to construct instances. Pair with
/// <see cref="ComponentVM{M}"/> when a typed model is required.
/// </summary>
public sealed class ComponentVM : ComponentVMBase, IComponentVM
{
    /// <inheritdoc/>
    public override ViewModelType Type => ViewModelType.Component;

    private ComponentVM(
        string name,
        string hint,
        IMessageHub hub,
        IDispatcher dispatcher,
        Action? onConstruct,
        Action? onDestruct,
        bool background)
        : base(name, hint, hub, dispatcher, onConstruct, onDestruct, background)
    {
    }

    // ── Builder factory ──────────────────────────────────────────────────────
    /// <summary>Returns a new empty builder for <see cref="ComponentVM"/>.</summary>
    public static ComponentVMBuilder Builder() => ComponentVMBuilder.Empty;

    // ── Options factory (additive; ADR-0055 / VMX-020) ──────────────────────
    /// <summary>
    /// Constructs a <see cref="ComponentVM"/> from a <see cref="ComponentVMOptions"/>
    /// record in a single call — an additive alternative to the fluent
    /// <see cref="ComponentVMBuilder"/>. Delegates to that builder, so the required-field
    /// validation (<see cref="BuilderValidationException"/> on a missing Name/Hub/Dispatcher)
    /// and the resulting VM are identical to the fluent path.
    /// </summary>
    public static ComponentVM Create(ComponentVMOptions options)
    {
        var b = ComponentVMBuilder.Empty
            .Hint(options.Hint)
            .Background(options.Background);
        if (options.Name is not null) b = b.Name(options.Name);
        if (options.Hub is not null) b = b.OptionHub(options.Hub);
        if (options.Dispatcher is not null) b = b.OptionDispatcher(options.Dispatcher);
        if (options.OnConstruct is not null) b = b.OnConstruct(options.OnConstruct);
        if (options.OnDestruct is not null) b = b.OnDestruct(options.OnDestruct);
        return b.Build();
    }

    // ── Internal factory used by builder ────────────────────────────────────
    internal static ComponentVM Create(
        string name,
        string hint,
        IMessageHub hub,
        IDispatcher dispatcher,
        Action? onConstruct,
        Action? onDestruct,
        bool background)
        => new(name, hint, hub, dispatcher, onConstruct, onDestruct, background);
}

/// <summary>
/// Sealed, modeled leaf viewmodel. Model is settable after construction.
///
/// See spec/05-component-vm.md §Variants (ComponentVM&lt;M&gt;, type=Component).
/// Use <c>ComponentVM&lt;M&gt;.Builder()</c> to construct instances.
/// </summary>
/// <typeparam name="M">The model type.</typeparam>
public sealed class ComponentVM<M> : ComponentVMBaseOfM<M>, IComponentVM<M>
{
    private readonly ViewModelType _type;

    /// <inheritdoc/>
    public override ViewModelType Type => _type;

    /// <inheritdoc/>
    public M Model
    {
        get => ModelValue;
        set => ModelValue = value;
    }

    private ComponentVM(
        string name,
        string hint,
        ViewModelType type,
        M model,
        Func<M, string> modeledHinter,
        Action<M>? onModelChanged,
        IMessageHub hub,
        IDispatcher dispatcher,
        Action? onConstruct,
        Action? onDestruct,
        bool background = false)
        : base(name, hint, model, modeledHinter, onModelChanged, hub, dispatcher, onConstruct, onDestruct, background)
    {
        _type = type;
    }

    // ── Builder factory ──────────────────────────────────────────────────────
    /// <summary>Returns a new empty builder for <see cref="ComponentVM{M}"/>.</summary>
    public static ComponentVMBuilder<M> Builder() => ComponentVMBuilder<M>.Empty;

    // ── Options factory (additive; ADR-0055 / VMX-020) ──────────────────────
    /// <summary>
    /// Constructs a <see cref="ComponentVM{M}"/> from a <see cref="ComponentVMOptions{M}"/>
    /// record in a single call — an additive alternative to the fluent
    /// <see cref="ComponentVMBuilder{M}"/>. Delegates to that builder, so the required-field
    /// validation (<see cref="BuilderValidationException"/> on a missing Name/Hub/Dispatcher)
    /// and the resulting VM are identical to the fluent path.
    /// </summary>
    public static ComponentVM<M> Create(ComponentVMOptions<M> options)
    {
        var b = ComponentVMBuilder<M>.Empty
            .Hint(options.Hint)
            .Type(options.Type)
            .Model(options.Model)
            .Background(options.Background);
        if (options.Name is not null) b = b.Name(options.Name);
        if (options.Hub is not null) b = b.OptionHub(options.Hub);
        if (options.Dispatcher is not null) b = b.OptionDispatcher(options.Dispatcher);
        if (options.ModeledHinter is not null) b = b.ModeledHinter(options.ModeledHinter);
        if (options.OnModelChanged is not null) b = b.OnModelChanged(options.OnModelChanged);
        if (options.OnConstruct is not null) b = b.OnConstruct(options.OnConstruct);
        if (options.OnDestruct is not null) b = b.OnDestruct(options.OnDestruct);
        return b.Build();
    }

    // ── Internal factory used by builder ────────────────────────────────────
    internal static ComponentVM<M> Create(
        string name,
        string hint,
        ViewModelType type,
        M model,
        Func<M, string> modeledHinter,
        Action<M>? onModelChanged,
        IMessageHub hub,
        IDispatcher dispatcher,
        Action? onConstruct,
        Action? onDestruct,
        bool background = false)
        => new(name, hint, type, model, modeledHinter, onModelChanged, hub, dispatcher, onConstruct, onDestruct, background);
}

File: langs/python/src/vmx/components/component_vm.py
"""ComponentVM (non-modeled) and ComponentVMOf[M] (modeled with settable model).

See spec/05-component-vm.md §Variants.
"""

from __future__ import annotations

from collections.abc import Callable
from typing import Generic, TypeVar

from vmx.components.base import _ComponentVMBase
from vmx.components.protocols import ViewModelType
from vmx.messages.protocols import Message
from vmx.services.dispatcher import Dispatcher
from vmx.services.message_hub import MessageHubProto

M = TypeVar("M")


# ---------------------------------------------------------------------------
# ComponentVM — non-modeled leaf VM
# ---------------------------------------------------------------------------


class ComponentVM(_ComponentVMBase):
    """Non-modeled leaf viewmodel.

    No model field. Use ``ComponentVM.builder()`` to construct instances.

    Type identifier: ``ViewModelType.COMPONENT``.
    """

    def __init__(
        self,
        *,
        name: str,
        hint: str,
        hub: MessageHubProto[Message],
        dispatcher: Dispatcher,
        on_construct: Callable[[], None] | None = None,
        on_destruct: Callable[[], None] | None = None,
        background: bool = False,
    ) -> None:
        super().__init__(
            name=name,
            hint=hint,
            hub=hub,
            dispatcher=dispatcher,
            on_construct=on_construct,
            on_destruct=on_destruct,
            background=background,
        )

    @property
    def type(self) -> ViewModelType:
        return ViewModelType.COMPONENT

    @staticmethod
    def builder() -> ComponentVMBuilder:
        """Return a new immutable builder for :class:`ComponentVM`."""
        return ComponentVMBuilder()

    @classmethod
    def create(
        cls,
        *,
        name: str | None = None,
        hub: MessageHubProto[Message] | None = None,
        dispatcher: Dispatcher | None = None,
        hint: str = "",
        on_construct: Callable[[], None] | None = None,
        on_destruct: Callable[[], None] | None = None,
        background: bool = False,
    ) -> ComponentVM:
        """Construct a :class:`ComponentVM` from keyword options in one call.

        An additive alternative to :meth:`builder` (ADR-0055 / VMX-020).
        Delegates to :class:`ComponentVMBuilder`, so required-field validation
        (``BuilderValidationError`` on a missing ``name``/``hub``/``dispatcher``)
        and the resulting VM are identical to the fluent path.
        """
        builder = ComponentVMBuilder().hint(hint).background(background)
        if name is not None:
            builder = builder.name(name)
        if hub is not None and dispatcher is not None:
            builder = builder.services(hub, dispatcher)
        if on_construct is not None:
            builder = builder.on_construct(on_construct)
        if on_destruct is not None:
            builder = builder.on_destruct(on_destruct)
        return builder.build()


# ---------------------------------------------------------------------------
# ComponentVMOf[M] — modeled leaf VM (settable model)
# ---------------------------------------------------------------------------


class ComponentVMOf(Generic[M], _ComponentVMBase):
    """Modeled leaf viewmodel with settable model.

    Use ``ComponentVMOf.builder()`` to construct instances.

    Type identifier: ``ViewModelType.COMPONENT``.
    """

    def __init__(
        self,
        *,
        name: str,
        hint: str,
        initial_model: M,
        modeled_hinter: Callable[[M], str],
        on_model_changed: Callable[[M], None] | None,
        hub: MessageHubProto[Message],
        dispatcher: Dispatcher,
        on_construct: Callable[[], None] | None = None,
        on_destruct: Callable[[], None] | None = None,
        background: bool = False,
        vm_type: ViewModelType = ViewModelType.COMPONENT,
    ) -> None:
        super().__init__(
            name=name,
            hint=hint,
            hub=hub,
            dispatcher=dispatcher,
            on_construct=on_construct,
            on_destruct=on_destruct,
            background=background,
        )
        self._model: M = initial_model
        self._modeled_hinter: Callable[[M], str] = modeled_hinter
        self._on_model_changed_cb: Callable[[M], None] | None = on_model_changed
        self._modeled_hint: str = modeled_hinter(initial_model)
        self._vm_type: ViewModelType = vm_type

    @property
    def type(self) -> ViewModelType:
        return self._vm_type

    @property
    def model(self) -> M:
        """The current model value."""
        return self._model

    @model.setter
    def model(self, value: M) -> None:
        """Set the model, emitting PropertyChangedMessage only on actual change."""
        self._set_model(value)

    @property
    def modeled_hint(self) -> str:
        """Derived hint computed from the current model via modeled_hinter."""
        return self._modeled_hint

    def _set_model(self, value: M) -> None:
        """Apply equality-guarded model update per spec/05-component-vm.md."""
        if self._is_disposed():
            return
        if self._model == value:
            return

        self._model = value

        self._notify_property_changed("model")

        # Recompute modeled_hint.
        new_hint = self._modeled_hinter(value)
        if self._modeled_hint != new_hint:
            self._modeled_hint = new_hint
            self._notify_property_changed("modeled_hint")

        # Invoke on_model_changed callback.
        if self._on_model_changed_cb is not None:
            self._on_model_changed_cb(value)

    @staticmethod
    def builder() -> ComponentVMOfBuilder[M]:
        """Return a new immutable builder for :class:`ComponentVMOf`."""
        return ComponentVMOfBuilder()

    @classmethod
    def create(
        cls,
        *,
        name: str | None = None,
        model: M,
        hub: MessageHubProto[Message] | None = None,
        dispatcher: Dispatcher | None = None,
        hint: str = "",
        modeled_hinter: Callable[[M], str] | None = None,
        on_model_changed: Callable[[M], None] | None = None,
        on_construct: Callable[[], None] | None = None,
        on_destruct: Callable[[], None] | None = None,
        background: bool = False,
        vm_type: ViewModelType = ViewModelType.COMPONENT,
    ) -> ComponentVMOf[M]:
        """Construct a :class:`ComponentVMOf` from keyword options in one call.

        An additive alternative to :meth:`builder` (ADR-0055 / VMX-020).
        Delegates to :class:`ComponentVMOfBuilder`, so required-field validation
        (``BuilderValidationError`` on a missing ``name``/``hub``/``dispatcher``)
        and the resulting VM are identical to the fluent path. ``model`` is a
        required keyword.
        """
        builder: ComponentVMOfBuilder[M] = ComponentVMOfBuilder()
        builder = builder.model(model).hint(hint).background(background).vm_type(vm_type)
        if name is not None:
            builder = builder.name(name)
        if hub is not None and dispatcher is not None:
            builder = builder.services(hub, dispatcher)
        if modeled_hinter is not None:
            builder = builder.modeled_hinter(modeled_hinter)
        if on_model_changed is not None:
            builder = builder.on_model_changed(on_model_changed)
        if on_construct is not None:
            builder = builder.on_construct(on_construct)
        if on_destruct is not None:
            builder = builder.on_destruct(on_destruct)
        return builder.build()


# Deferred import to avoid circular references — builders imports component_vm.
from vmx.components.builders import ComponentVMBuilder, ComponentVMOfBuilder  # noqa: E402
